vortex-cxx/cpp/src/scan.cpp

```diff
@@ -4,33 +4,96 @@
 #include "vortex/scan.hpp"
 #include "vortex/exception.hpp"
 #include "rust/cxx.h"
+#include "vortex/expr.hpp"
 
 namespace vortex {
-ScanBuilder &&ScanBuilder::WithFilter(Expr expr) && {
-    impl_->with_filter(std::move(expr).IntoImpl());
+
+template <typename ExprType>
+ScanBuilder &ScanBuilder::WithFilter(ExprType &&expr) & {
+    if constexpr (std::is_rvalue_reference_v<ExprType &&>) {
+        impl_->with_filter(std::forward<ExprType>(expr).IntoImpl());
+    } else {
+        impl_->with_filter_ref(std::forward<ExprType>(expr).Impl());
+    }
+    return *this;
+}
+template ScanBuilder &ScanBuilder::WithFilter(expr::Expr &&expr) &;
+template ScanBuilder &ScanBuilder::WithFilter(const expr::Expr &expr) &;
+
+template <typename ExprType>
+ScanBuilder &&ScanBuilder::WithFilter(ExprType &&expr) && {
+    if constexpr (std::is_rvalue_reference_v<ExprType &&>) {
+        impl_->with_filter(std::forward<ExprType>(expr).IntoImpl());
+    } else {
+        impl_->with_filter_ref(std::forward<ExprType>(expr).Impl());
+    }
```

Comment:
cxx does not support function overload (although I think technically it can do it), so we have to add separate ffi functions and do static dispatch here.

Reply:
I think the forwarding part is less important than the "reference collapsing" aspect such that the builder function can handle both rvalues and lvalues. Let's drop `std::forward` and a keep a single `with_xyz` not adding a ref version.

Reply:
If there are only two variants, and we still need specialization for both of them (to call the `with_filter_ref` ffi), then we do not need forward I think

Reply:
Would you mind taking a look at the modifications of my latest commit?  I think it is cleaner

Reply:
I'll have a second thought here also on whether C++23 might be ok to use so we could express this with reference collapsing and auto-deducing `this`.

Reply:
I think C++ 20 may be ok but C++23 is too aggressive. As I know someone is still using C++17 (e.g., my current employer). Arrow-cpp recently had a [vote](https://lists.apache.org/thread/48zlj0dn2y0f53y2k37qsr90y781wfnj) to switch to 20 but it has not happened yet.